File: src/com/mistraltech/smogen/codegenerator/Generator.java
package com.mistraltech.smogen.codegenerator;

import com.intellij.codeInsight.actions.AbstractLayoutCodeProcessor;
import com.intellij.codeInsight.actions.OptimizeImportsProcessor;
import com.intellij.codeInsight.actions.RearrangeCodeProcessor;
import com.intellij.codeInsight.actions.ReformatCodeProcessor;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.JavaPsiFacade;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static com.mistraltech.smogen.utils.ActionUtils.runAction;
import static com.mistraltech.smogen.utils.PsiUtils.createMissingDirectoriesForPackage;
import static com.mistraltech.smogen.utils.PsiUtils.findDirectoryForPackage;

public class Generator {

    private final GeneratorProperties generatorProperties;

    public Generator(@NotNull final GeneratorProperties generatorProperties) {
        this.generatorProperties = generatorProperties;
    }

    public void generate() {
        final PsiDirectory parentDirectory = findOrCreateParentDirectory();
        generatorProperties.setParentDirectory(parentDirectory);

        final PsiFile existingFile = parentDirectory.findFile(generatorProperties.getFileName());
        if (existingFile != null && !shouldOverwriteFile(existingFile)) {
            return;
        }

        List<String> warnings = new ArrayList<String>();
        warnings.addAll(checkProperties());
        if (! warnings.isEmpty() && !shouldContinueWithWarnings(warnings)) {
            return;
        }

        final PsiFile generatedFile = generateFile();

        reformat(generatedFile);
    }

    protected List<String> checkProperties() {
        return Collections.emptyList();
    }

    private PsiFile generateFile() {
        FileContentGeneratorRunnable fileContentGeneratorRunnable = new FileContentGeneratorRunnable(generatorProperties);

        runAction(fileContentGeneratorRunnable);

        return fileContentGeneratorRunnable.getTargetFile();
    }

    private void reformat(PsiFile targetFile) {
        AbstractLayoutCodeProcessor processor = new ReformatCodeProcessor(targetFile.getProject(), targetFile, null, false);
        processor = new OptimizeImportsProcessor(processor);
        processor = new RearrangeCodeProcessor(processor);
        processor.run();
    }

    protected boolean shouldOverwriteFile(PsiFile existingFile) {
        String msg = "File " + existingFile.getVirtualFile().getPresentableUrl() + " already exists. Overwrite?";
        return Messages.showYesNoDialog(generatorProperties.getProject(), msg, "File Already Exists",
                "Overwrite", "Cancel", Messages.getWarningIcon()) == Messages.YES;
    }

    protected boolean shouldContinueWithWarnings(List<String> warnings)
    {
        String msg = "The generated source may not compile for the following reasons:\n" + StringUtils.join(warnings, '\n');
        return Messages.showYesNoDialog(generatorProperties.getProject(), msg, "Generated Code Will Not Compile",
                "Continue", "Cancel", Messages.getWarningIcon()) == Messages.YES;
    }

    @NotNull
    private PsiDirectory findOrCreateParentDirectory() {
        PsiManager psiManager = PsiManager.getInstance(generatorProperties.getProject());

        VirtualFile sourceRoot = generatorProperties.getSourceRoot();

        final PsiDirectory baseDir = Optional.ofNullable(psiManager.findDirectory(sourceRoot))
                .orElseThrow(() -> new IllegalStateException("Expected source root directory to exist"));

        // Ensure that the directory for the target package exists
        ApplicationManager.getApplication().runWriteAction(() -> {
            createMissingDirectoriesForPackage(baseDir, generatorProperties.getPackageName());
        });

        PsiDirectory directory = findDirectoryForPackage(baseDir, generatorProperties.getPackageName())
                .orElseThrow(() -> new IllegalStateException("Expected directory for package to have been created"));

        return directory;
    }

    private static class FileContentGeneratorRunnable implements Runnable {
        private final GeneratorProperties generatorProperties;
        private PsiFile targetFile;

        FileContentGeneratorRunnable(final GeneratorProperties generatorProperties) {
            this.generatorProperties = generatorProperties;
        }

        @Override
        public void run() {
            final PsiDirectory parentDirectory = generatorProperties.getParentDirectory();
            final String targetFileName = generatorProperties.getFileName();
            final PsiFile existingFile = parentDirectory.findFile(targetFileName);
            targetFile = existingFile != null ? existingFile : parentDirectory.createFile(targetFileName);

            PsiDocumentManager documentManager = PsiDocumentManager.getInstance(parentDirectory.getProject());
            Document document = Optional.ofNullable(documentManager.getDocument(targetFile))
                    .orElseThrow(() -> new IllegalStateException("Expected document from file " + targetFile.getName()));

            document.setText(generatorProperties.getCodeWriter().writeCode());

            documentManager.commitDocument(document);
        }

        /**
         * Gets the file containing the newly generated matcher class.
         *
         * @return the matcher class file
         */
        PsiFile getTargetFile() {
            return targetFile;
        }
    }
}

File: src/main/kotlin/com/mistraltech/smogen/codegenerator/Generator.kt
package com.mistraltech.smogen.codegenerator

import com.intellij.codeInsight.actions.AbstractLayoutCodeProcessor
import com.intellij.codeInsight.actions.OptimizeImportsProcessor
import com.intellij.codeInsight.actions.RearrangeCodeProcessor
import com.intellij.codeInsight.actions.ReformatCodeProcessor
import com.intellij.openapi.ui.Messages
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiDocumentManager
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiManager
import com.mistraltech.smogen.utils.ActionUtils.runAction
import com.mistraltech.smogen.utils.PsiUtils.createMissingDirectoriesForPackage
import com.mistraltech.smogen.utils.PsiUtils.findDirectoryForPackage

open class Generator(private val generatorProperties: GeneratorProperties<*>) {

    fun generate() {
        val parentDirectory = findOrCreateParentDirectory()
        generatorProperties.setParentDirectory(parentDirectory)

        val existingFile = parentDirectory.findFile(generatorProperties.fileName)
        if (existingFile != null && !shouldOverwriteFile(existingFile)) {
            return
        }

        val warnings = checkProperties()
        if (warnings.isNotEmpty() && !shouldContinueWithWarnings(warnings)) {
            return
        }

        val generatedFile = generateFile()

        reformat(generatedFile)
    }

    protected open fun checkProperties(): List<String> {
        return emptyList()
    }

    private fun generateFile(): PsiFile {
        val fileContentGeneratorRunnable = FileContentGeneratorRunnable(generatorProperties)

        runAction(fileContentGeneratorRunnable)

        return fileContentGeneratorRunnable.targetFile!!
    }

    private fun reformat(targetFile: PsiFile) {
        var processor: AbstractLayoutCodeProcessor = ReformatCodeProcessor(targetFile.project, targetFile, null, false)
        processor = OptimizeImportsProcessor(processor)
        processor = RearrangeCodeProcessor(processor)
        processor.run()
    }

    protected open fun shouldOverwriteFile(existingFile: PsiFile): Boolean {
        return Messages.showYesNoDialog(
            generatorProperties.project,
            "File " + existingFile.virtualFile.presentableUrl + " already exists. Overwrite?",
            "File Already Exists",
            "Overwrite",
            "Cancel",
            Messages.getWarningIcon()
        ) == Messages.YES
    }

    protected open fun shouldContinueWithWarnings(warnings: List<String>): Boolean {
        return Messages.showYesNoDialog(
            generatorProperties.project,
            "The generated source may not compile for the following reasons:\n" +
                (warnings.joinToString("\n")),
            "Generated Code Will Not Compile",
            "Continue",
            "Cancel",
            Messages.getWarningIcon()
        ) == Messages.YES
    }

    private fun findOrCreateParentDirectory(): PsiDirectory {
        val psiManager = PsiManager.getInstance(generatorProperties.project!!)
        val sourceRoot = generatorProperties.sourceRoot
        val baseDir = psiManager.findDirectory(sourceRoot!!)
            ?: throw IllegalStateException("Expected source root directory to exist")

        // Ensure that the directory for the target package exists
        runAction { createMissingDirectoriesForPackage(baseDir, generatorProperties.packageName!!) }

        return findDirectoryForPackage(baseDir, generatorProperties.packageName!!)
            ?: throw IllegalStateException("Expected directory for package to have been created")
    }

    private class FileContentGeneratorRunnable(private val generatorProperties: GeneratorProperties<*>) : Runnable {
        /**
         * Gets the file containing the newly generated matcher class.
         *
         * @return the matcher class file
         */
        var targetFile: PsiFile? = null
            private set

        override fun run() {
            val parentDirectory = generatorProperties.parentDirectory
            val targetFileName = generatorProperties.fileName

            val existingFile = parentDirectory!!.findFile(targetFileName)

            targetFile = existingFile ?: parentDirectory.createFile(targetFileName)

            val documentManager = PsiDocumentManager.getInstance(parentDirectory.project)
            val document = documentManager.getDocument(targetFile!!)
                ?: throw IllegalStateException("Expected document from file " + targetFile!!.name)

            document.setText(generatorProperties.codeWriter.writeCode())
            documentManager.commitDocument(document)
        }
    }
}
